ldap/servers/slapd/connection.c

```diff
@@ -271,12 +271,18 @@ connection_reset(Connection *conn, int ns, PRNetAddr *from, int fromLen __attrib
     char buf_destip[INET6_ADDRSTRLEN + 1] = {0};
     char *str_unknown = "unknown";
     int in_referral_mode = config_check_referral_mode();
+    int proxy_connection = 0;
+    PRNetAddr pr_netaddr_dest = {{0}};
 
     slapi_log_err(SLAPI_LOG_CONNS, "connection_reset", "new %sconnection on %d\n", pTmp, conn->c_sd);
 
     /* bump our count of connections and update SNMP stats */
     conn->c_connid = slapi_counter_increment(num_conns);
 
+    if (haproxy_receive(conn->c_sd, &proxy_connection, from, &pr_netaddr_dest) != 0) {
+        slapi_log_err(SLAPI_LOG_CONNS, "connection_reset", "Failed to recieve HAProxy information\n");
+    }
+
```

Comment:
Several things are wrong here:
  - if haproxy_receive returns that an header is found.
      - IMPORTANT FOR SECURITY: Should check that the socket sender address (i.e the proxy host) is allowed to 
        send a proxy request (i.e  a new config attribute is needed)
      The pr_netaddr_dest (i.e the proxyed host) should be stored in the connection (so that aci/ host check - check the original host  rather than the proxy host )

Reply:
I do not know if this is the same concern as @progier389 said, but in case haproxy_receive can be broken by a craft message we should have a config mechanism to disable the call to haproxy_receive.

Reply:
Yes I was thinking about a multivalued list of trusted IP addresses. I do not think that we can afford not to do an access control at this level: 
 Firewall cannot help the deployment that does not use HAProxy because in that case any client that have access to the server could then fake its address by crafting an HAProxy header
 
Furthermore the HAProxy reference document explicitly warn against that in its security consideration section:
> Implementers should be very careful about not trying to automatically detect
> whether they have to decode the header or not, but rather they must only rely
> on a configuration parameter. Indeed, if the opportunity is left to a normal
> client to use the protocol, he will be able to hide his activities or make them
> appear as coming from someone else. However, accepting the header only from a
> number of known sources should be safe.

 I do not think that we should  wholly disable the call to haproxy_receive as @tbordaz suggest I think that decoding the haproxy record and rejecting it if it is not allowed is more clear in term of diagnostic 
  (You could explicitly see in access log that there was an HAProxy attempt and that it was rejected)
 
 Another point: IMHO we should log a message in access log that the connection is proxyed (mentionning the proxied  address) (Because otherwise we may have problem with troubleshooting ... )
 I was thinking about:
  [date] conn=N HAPROXY version=1 new_address=.... OK/denied

Reply:
Maybe a log message to the security logs also ?

Reply:
I agreed with @progier389 here, we need a way to say that we accept and trust the HAPROXY info we are being sent. If that isn't set, then we should reject connections that attempt to tamper with this.